### PR TITLE
Selenium 4 Alpha

### PR DESCRIPTION
**Pull request by tcnh**
This PR contains a whole bunch of things:
 * Selenium 4 Alpha 6, allowing for:
    * ChromiumDrivers/Options usage. Which is implemented in LocalDriverFactory. Some examples of chromium (Chrome/Modern Edge configuration) are in the wiki. Should make configuration more straightforward.
    * Relative Locators - Click/Enter as methods as examples are in BrowserTest. (i.e. `| click | input | left of | bad coded checkbox label |`
    * DevTools API for chromium (examples in the wiki files). I.E:
      * Request Header injection
      * Basic/Digest authentication
      * Cookie injection
      * Geo Location Mocking
      * (XHR) Response overrides/mocking
      * Request Logging
      * Performance Metrics
      * Full page screen shots

This is not production-ready (the alpha status of Selenium 4 with all of its quirks alone proves that), but I'd like to see some review/testing it in a bit of a broader perspective than just my own. I believe the possibilities (especially leveraging devTools) can prove valuable.

### CODE REVIEW

**Comment by fhoeben** [comment]
I tried the code in [a GitLab pipeline](https://fhoeben.gitlab.io/-/hsac-fitnesse-fixtures/-/jobs/813635919/artifacts/full-results/index.html) (using chrome in docker) and I see the following message. Since I am using chrome the message is not completely right. Can this code only be used with a local chrome, or do I need to upgrade to a Selenium 4 grid/docker image, or is the test performed not completely correct, or ...?

> nl.hsac.fitnesse.fixture.slim.StopTestException: message:<<DevTools enabled Browser Test can only be used with a chromium based browser (Chrome/Edge)>>

**Comment by fhoeben** [comment]
I now see that there are chrome standalone images for selenium 4 (e.g. selenium/standalone-chrome:4.0.0-alpha-7-prerelease-20201009). Do we need to upgrade to one of those?

**Comment by tcnh** [comment]
I believe that when using grid (or docker images), the grid version should be in sync with the client version to prevent unexpected behavior. I have to admit I didn't have a chance to test with grid yet

**Comment by fhoeben** [comment]
I don't believe you can use the devtools with Selenium grid. I found a couple of questions about it on stack overflow, but no answers. But with a different setup of how I use chrome in docker it does work. It doesn't seem to run the same tests as stable as Selenium 4 yet (see the Gitlab runs).

I also added some very minor code cleanup

**Comment by tcnh** [comment]
Upgraded to alpha 7 (came out yesterday). Removed overridden RelativeLocator class, some refactoring in DevTools (still needs work)

File: pom.xml [hunk 1]
```diff
@@ -403,7 +384,7 @@
 					<properties>
 						<property>
 							<name>listener</name>
-							<value>fitnesse.junit.JUnitXMLPerPageRunListener,nl.hsac.fitnesse.junit.ProgressLoggerListener${extraFailsafeListeners}</value>
+							<value>nl.hsac.fitnesse.junit.JUnitXMLPerPageListener,nl.hsac.fitnesse.junit.ProgressLoggerListener${extraFailsafeListeners}</value>
```
Comment (fhoeben) [inline pom.xml:387]: Is this change intentional?

**Review by fhoeben — COMMENTED** [review]
Seems to work quite good, all acceptance test pass.

Could you look at adding some acceptance tests for the new fixtures and new Selenium 4 features (such as dev tools usage)?

**Comment by fhoeben** [comment]
On my [gitlab run](https://gitlab.com/fhoeben/hsac-fitnesse-fixtures/-/pipelines/403522776) I did have [a job](https://gitlab.com/fhoeben/hsac-fitnesse-fixtures/-/jobs/1755706856) that seems to hang, see partial output below.

Other example runs did not run flawlessly either, but unfortunately this is also the case on the master branch.


```
2021-11-06 14:49:02.772 test Started 'HsacExamples.SlimTests.BrowserTests.Selenium4Examples.BrowserTestWithDevTools.InjectCookie' (4 / 6)
Nov 06, 2021 2:49:09 PM org.openqa.selenium.devtools.Connection$Listener lambda$onText$0
WARNING: Unable to process: {"method":"Fetch.requestPaused","params":{"requestId":"interception-job-45.0","request":{"url":"https://s7.addthis.com/static/sh.f48a1a04fe8dbf021b4cda1d.html","urlFragment":"#rand=0.52682044711052&iit=1636210149209&tmr=load%3D1636210148988%26core%3D1636210149201%26main%3D1636210149202%26ifr%3D1636210149215&cb=0&cdn=0&md=0&kw=&ab=-&dh=orangebeard.io&dr=&du=https%3A%2F%2Forangebeard.io%2F&href=https%3A%2F%2Forangebeard.io%2F&dt=Orangebeard%20%7C%20Test%20software%20like%20a%20boss.&dbg=0&cap=tc%3D0%26ab%3D0&inst=1&jsl=1&prod=undefined&lng=en&ogt=alt%2Cheight%2Cwidth%2Csecure_url%2Cimage%2Cdescription%2Ctitle%2Ctype%3Dwebsite&pc=wpf&pub=wp-a202a4a71008feb61e28c160e62d8990&ssl=1&sid=618695e5a404bac7&srf=0.01&ver=300&xck=0&xtr=0&og=url%3Dhttps%253A%252F%252Forangebeard.io%252F%26site_name%3DOrangebeard%26locale%3Den_US%26type%3Dwebsite%26title%3DOrangebeard%2520%257C%2520Test%2520software%2520like%2520a%2520boss.%26description%3DAn%2520intelligent%2520dashboard%2520based%2520platform%252C%2520showing%2520you%2520real%2520time%2520results%2520of%2520all%2520automated%2520tests%2520(of%2520any%2520kind!)%2520throughout%2520your%2520test%2520pyramid.%26image%3Dhttps%253A%252F%252Forangebeard.io%252Fwp-content%252Fuploads%252F2020%252F08%252FSchermafbeelding-2020-07-17-om-11.36.22.png%26secure_url%3Dhttps%253A%252F%252Forangebeard.io%252Fwp-content%252Fuploads%252F2020%252F08%252FSchermafbeelding-2020-07-17-om-11.36.22.png%26width%3D2316%26height%3D1138%26alt%3DOrangebeard%2520logo&csi=undefined&rev=v8.28.8-wp&ct=1&xld=1&xd=1","method":"GET","headers":{"sec-ch-ua":"\"Google Chrome\";v=\"95\", \"Chromium\";v=\"95\", \";Not A Brand\";v=\"99\"","sec-ch-ua-mobile":"?0","sec-ch-ua-platform":"\"Linux\"","Upgrade-Insecure-Requests":"1","User-Agent":"Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/95.0.4638.69 Safari/537.36","Accept":"text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.9","Referer":"https://orangebeard.io/"},"initialPriority":"VeryHigh","referrerPolicy":"no-referrer-when-downgrade"},"frameId":"C3E5E944EEB63B40FD048B935EA9778B","resourceType":"Document","networkId":"776AEE7DD14657C280C9BD91026628AB"},"sessionId":"762671517E5ECE0F295E2A8D56FAF40F"}
org.openqa.selenium.devtools.DevToolsException: {"id":119,"error":{"code":-32602,"message":"Invalid InterceptionId."},"sessionId":"762671517E5ECE0F295E2A8D56FAF40F"}
Build info: version: '4.0.0', revision: '3a21814679'
System info: host: 'runner-ed2dce3a-project-17353740-concurrent-0', ip: '172.17.0.3', os.name: 'Linux', os.arch: 'amd64', os.version: '5.4.109+', java.version: '11.0.11'
Driver info: driver.version: unknown
	at org.openqa.selenium.devtools.Connection.sendAndWait(Connection.java:159)
...
```

**Comment by rvanderboom** [comment]
is it possible to test this pre-release, can i put a snapshot version in my fitnesse project pom ? Or a jar incan download ?
would love to try it in our test suite

**Comment by fhoeben** [comment]
@tcnh have you had any chance to look at this further?

**Comment by tcnh** [comment]
Not yet, I expect to be able to spare some time between Christmas and new
year..

Op vr 17 dec. 2021 14:32 schreef Fried Hoeben ***@***.***>:

**Comment by fhoeben** [comment]
@tcnh, best wishes for the new year. Did you get around to finishing this PR?

**Comment by tcnh** [comment]
You too! Unfortunately I've been out of the running for more than a week due to a covid infection that was a bit more severe than the runny nose I had expected. It's on my to do list!

**Comment by tcnh** [comment]
Closing thisin favor of a new PR

**Pull request closed without merging**